### PR TITLE
Fix namespace switch issues and add ipvlan, macvlan, and bridge e2e testing

### LINKED ISSUE #179: {mac,ip}vlan and bridge plugins need tests

### PR DESCRIPTION
**Pull request by dcbw**
A subset of https://github.com/containernetworking/cni/pull/167 while we discuss how the mocking should actually work.

Fixes #179.

### CODE REVIEW

**Comment by dcbw** [comment]
cc @rosenhouse

**Comment by dcbw** [comment]
cc @steveeJ

**Comment by dcbw** [comment]
@rosenhouse one thing I'm not 100% sure about is the removal of LockOSThread() from the plugins init() functions themselves; that all works for the testcases because they run all the checks with Do(), and they run the plugin's cmdAdd/cmdDel from inside the Do() for the original namespace too.  But when the plugins are run standalone from kubernetes or rkt or whatever, should they get the current host namespace and wrap inital setup with Do()?

**Comment by steveeJ** [comment]
> But when the plugins are run standalone from kubernetes or rkt or whatever, should they get the current host namespace and wrap inital setup with Do()?
AFAIU this shouldn't be a problem, reasoning below.

If the plugins are invoked, the given network namespace will be explicitly set whenever necessary using `NetNS.Do()`. AFAICT the assumption is that the rest of the code is supposed to run in the caller's netns, and unless the caller doesn't mess with setting the netns themselves, CNI shouldn't be messing with it either.

**Comment by dcbw** [comment]
@steveeJ yeah, that should be true.  I guess the testcases have issues here (which were fixed by wrapping everything in Do()) because they create a completely new namespace to ensure they don't affect the host.  So I guess we're OK.

**Comment by steveeJ** [comment]
> yeah, that should be true. I guess the testcases have issues here (which were fixed by wrapping everything in Do()) because they create a completely new namespace to ensure they don't affect the host. So I guess we're OK.

Agreed! At this point, thank you again for this excellent contribution. I lack the overview if all of @rosenhouse's issues related to the code are addressed, @dcbw can you help here?

**Comment by dcbw** [comment]
rebased to resolve the merge conflict

**Comment by dcbw** [comment]
@steveeJ @rosenhouse I believe they were all addressed in the previous mocking PR; the PR here doesn't change anything from that other PR except for dropping the 'netops' object.  Should get @rosenhouse to confirm though.

**Comment by steveeJ** [comment]
As discussed in the CNI meeting, this will be merged after bringing the locking back to the plugins `init()` functions.

**Comment by dcbw** [comment]
@steveeJ reverted the init() changes; PTAL.

**Comment by rosenhouse** [comment]
So, like I said at the CNI meeting, I'm traveling on PTO, and then at a conference and therefore unable to look at this closely until late next week.

However, at first glance, I think I see instances of the issue I mentioned on the call: given how we're trying to isolate namespace switches to a goroutine, nested calls to `netns.Do` are problematic.  Specifically, you've got test code which calls `netns.Do` and then inside its callback, you directly invoke the plugin code.  And inside the plugin code (e.g. in the bridge plugin) there is another call to `netns.Do`.  That is likely the cause of the test flakes you're seeing.  Calling `LockOSThread` at the start of the test suite, as I suggested on the call, won't help here.

I think the easiest way to fix those flakes would be to invoke the plugins as a separate OS process, rather than running them in the same process as the tests.

I'll leave it up to the other maintainers to decide on merging.

**Comment by steveeJ** [comment]
> reverted the init() changes; PTAL.

I've checked the test cases before the revert, so I trust the green lights with this now.

**Comment by steveeJ** [comment]
@rosenhouse 
> Specifically, you've got test code which calls netns.Do and then inside its callback, you directly invoke the plugin code. And inside the plugin code (e.g. in the bridge plugin) there is another call to netns.Do. That is likely the cause of the test flakes you're seeing. Calling LockOSThread at the start of the test suite, as I suggested on the call, won't help here.

Let me try to repeat what you've explained to confirm and also lay the information open on the issue.
The first `Do()`'s goroutine is locked correctly, but in spite of this the nested goroutine could potentially be scheduled on a new OS thread because `GOMAXPROCS` has not been exhausted yet?

**Comment by rosenhouse** [comment]
@steveeJ exactly -- that new goroutine may or may not end up on a new OS thread.  If a new OS thread is created, then it will inherit the namespaces of its parent.  So the new OS thread's "original" namespace is actually the "switched-to" special namespace of the first call.  From that point on, the newly created OS thread is "polluted".  

I've updated https://github.com/containernetworking/cni/issues/183 to try to describe this issue and recommend against nested calls to this type of namespace switching code.  It's still not clear to me how to best expose namespace switching at a higher level of abstraction in a safe way.

**Comment by steveeJ** [comment]
> If a new OS thread is created, then it will inherit the namespaces of its parent. So the new OS thread's "original" namespace is actually the "switched-to" special namespace of the first call. From that point on, the newly created OS thread is "polluted".

This seems to go very deep down to Go's thread/routine handling mechanism and I would appreciate your source to your knowledge.

**Comment by rosenhouse** [comment]
@steveeJ I don't have a written reference for the fact that threads inherit the namespace context of their parent thread.  However, another member of our CloudFoundry container networking team (@sykesm) has investigated this issue.  I can share the results of his work here.

He wrote code that switched the namespace of the main goroutine and spun up several more goroutines from there.  All the goroutines waited while the program reported the namespace associated with each OS thread.  ([code here](https://gist.github.com/sykesm/020a27341cca5169250990d250b25db4))  Here are the results:
```
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15832/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15833/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15834/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15835/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15836/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15838/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15839/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15840/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15841/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15842/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15843/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15844/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15845/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15846/ns/net -> net:[4026533940]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15848/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/task/15849/ns/net -> net:[4026533719]
lrwxrwxrwx 1 root root 0 May  9 16:12 /proc/15832/ns/net -> net:[4026533719]
```
As you can see, the first few threads all share the original netns of the process, but several others are on the new netns.

As I mentioned earlier, I don't have time to dig into this now.  But it might be interesting to perform a more detailed investigation to show how this state of the world evolves, perhaps reporting the full set of `thread:netns` pairings at each step of the loop.

File: pkg/ns/ns.go [hunk 1]
```diff
@@ -15,64 +15,209 @@
...
 
-// SetNS sets the network namespace on a target file.
-func SetNS(f *os.File, flags uintptr) error {
-	_, _, err := syscall.RawSyscall(setNsNr, f.Fd(), flags, 0)
-	if err != 0 {
-		return err
-	}
+type NetNS interface {
+	// Executes the passed closure in this object's network namespace,
+	// restoring the original namespace afterwards. If the closure returns
```
Comment (steveeJ) [inline pkg/ns/ns.go]: as discussed OOB, we can't guarantee that the original namespace is restored. please add an explanation and maybe link to further docs, e.g. https://github.com/golang/go/wiki/LockOSThread

File: pkg/ns/README.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+### Namespaces, Threads, and Go
+On Linux each OS thread can have a different network namespace.  Go's thread scheduling model switches goroutines between OS threads based on OS thread load and whether the goroutine would block other goroutines.  This can result in a goroutine switching network namespaces without notice and lead to errors in your code.
+
+### Namespace Switching
+Switching namespaces with the `ns.Set()` method is not recommended without additional strategies to prevent unexpected namespace changes when your goroutines switch OS threads.
+
+Go provides the `runtime.LockOSThread()` function to ensure a specific goroutine executes on its current OS thread and prevents any other goroutine from running in that thread until the locked one exits.  Careful usage of `LockOSThread()` and short goroutines can provide good control over which network namespace a given goroutine executes in.
```
Comment (steveeJ) [inline pkg/ns/README.md]: "short goroutines" worries me a little, is that really necessary? A lock's purpose is to synchronize concurrent tasks and prevent any unwanted runtime behavior.

File: pkg/ns/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+For example, you cannot rely on the `ns.Set()` namespace being the current namespace after the `Set()` call unless the goroutine calling `Set()` has previously called `LockOSThread()`, and you ensure `runtime.UnlockOSThread()` is not called somewhere in-between.  You also cannot rely on the initial network namespace remaining the current network namespace of a goroutine if any other code in your program switches namespaces, unless you have already called `LockOSThread()` in that goroutine.  Note that `LockOSThread()` prevents the Go scheduler from optimally scheduling goroutines for best performance, so `LockOSThread()` should only be used in small, isolated goroutines that release the lock quickly.
```
Comment (steveeJ) [inline pkg/ns/README.md]: second sentence is too long IMO, do you mind splitting it apart?

File: pkg/ns/README.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+### Do() The Recommended Thing
+The `ns.Do()` method provides control over network namespaces for you by implementing these strategies. All code dependent on a particular network namespace should be wrapped in the `ns.Do()` method to ensure the correct namespace is selected for the duration of your code.
```
Comment (steveeJ) [inline pkg/ns/README.md]: shall we mention the tests as an example for usage? or should we give a short example here?

File: pkg/ns/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+For example, you cannot rely on the `ns.Set()` namespace being the current namespace after the `Set()` call unless you do two things.  First, the goroutine calling `Set()` must have previously called `LockOSThread()`.  Second, you must ensure `runtime.UnlockOSThread()` is not called somewhere in-between.  You also cannot rely on the initial network namespace remaining the current network namespace if any other code in your program switches namespaces, unless you have already called `LockOSThread()` in that goroutine.  Note that `LockOSThread()` prevents the Go scheduler from optimally scheduling goroutines for best performance, so `LockOSThread()` should only be used in small, isolated goroutines that release the lock quickly.
+
+### Do() The Recommended Thing
+The `ns.Do()` method provides control over network namespaces for you by implementing these strategies. All code dependent on a particular network namespace should be wrapped in the `ns.Do()` method to ensure the correct namespace is selected for the duration of your code.  For example:
+
```
Comment (steveeJ) [inline pkg/ns/README.md]: please wrap the code snippet in ```

**Pull request merged**